easybuild/easyconfigs/c/Cellformer/Cellformer-20240917-foss-2023a-R-4.3.2.eb

```diff
@@ -0,0 +1,190 @@
+easyblock = 'Tarball'
+
+name = 'Cellformer'
+version = '20240917'
+versionsuffix = '-R-%(rver)s'
+local_commit = '99a1165'
+
+homepage = 'https://github.com/elo-nsrb/Cellformer'
+description = '''An implementation of Cellformer from our publication: Berson et al.
+ "Whole genome deconvolution unveils Alzheimer’s resilient epigenetic signature"'''
+
+toolchain = {'name': 'foss', 'version': '2023a'}
+
+source_urls = ['https://github.com/elo-nsrb/Cellformer/archive/']
+sources = [{'download_filename': '%s.tar.gz' % local_commit, 'filename': '%(name)s-%(version)s.tar.gz'}]
+checksums = ['7cbddad75a4d47dfc0a39cd660ef20fe4e3cb755631b1b96136c1c3d5226c914']
+
+dependencies = [
+    ('Python', '3.11.3'),
+    ('PyTorch', '2.1.2'),
+    ('PyTorch-bundle', '2.1.2'),
+    ('scikit-learn', '1.3.1'),
+    ('PyTorch-Lightning', '2.2.1'),
+    ('anndata', '0.10.5.post1'),
+    ('h5py', '3.9.0'),
+    ('SciPy-bundle', '2023.07'),
+    ('Seaborn', '0.13.2'),
+    ('tensorboard', '2.15.1'),
+    ('tensorboardX', '2.6.2.2'),
+    ('torchvision', '0.16.0'),
+    ('tqdm', '4.66.1'),
+    ('scanpy', '1.9.8'),
+    ('pretty-yaml', '24.7.0'),
+    ('Arrow', '14.0.1'),
+    ('R', '4.3.2'),
+    ('R-bundle-CRAN', '2023.12'),
+    ('R-bundle-Bioconductor', '3.18', versionsuffix),
+    ('ArchR', '1.0.2', versionsuffix),
+    ('typing-extensions', '4.9.0'),
+    ('einops', '0.7.0'),
+]
+
+exts_defaultclass = 'PythonPackage'
+exts_default_options = {
+    'source_urls': [PYPI_SOURCE],
+    'download_dep_fail': True,
+    'use_pip': True,
+    'sanity_pip_check': True,
+    'installopts': '',
+}
+
+exts_list = [
+    ('typing_extensions', '4.12.2', {
```

Comment:
There should be no need for this, since `typing-extensions` is already a dependency?

Reply:
@PetrKralCZ Just for sanity's sake: in your test environment, the `scanpy/1.9.8-foss-2023a` module is still loading `hatchling/1.18.0-GCCcore-12.3.0` as a runtime dependency, which includes `typing-extensions` 4.6.3 as an extension. That's wrong, `hatchling` should only ever be a build dependency...

The same goes for the `scanpy` module in my test environment, but there's another difference that affects the order of `$PYTHONPATH`: in my test environment the `pydantic/1.10.13-GCCcore-12.3.0` module is old, it was not rebuilt after the changes in https://github.com/easybuilders/easybuild-easyconfigs/pull/18948/commits/479ecec53a1ac285392b1072181d7ea256d76fa8.

If `pydantic/1.10.13-GCCcore-12.3.0` loads `typing-extensions/4.9.0-GCCcore-12.3.0` (like it is in your case), then it gets superseded by the older `typing-extensions` that is included in the `hatchling` installation that gets loaded as a dependency of `scanpy`.

Long story short: we should make `hatchling` a *build* dependency of `scanpy`, and then the problem should be resolved... => being fixed via #21657

Reply:
Other then `scanpy` I also had to rebuild `anndata` on Donphan to pass the pip checks without `typing-extensions` in `exts_list`.